site/content/entry/improvements-to-hardware-renderer.md

```diff
@@ -0,0 +1,200 @@
+The first video is showing off the software renderer, which emulates the PICA200 by using your computer's CPU.
+On the other hand, the second video is using the OpenGL hardware renderer, which emulates the PICA200 by using your computer's GPU.
+From those videos, using your GPU to emulate the 3DS GPU is the clear winner when it comes to speed!
+However, it's not all sunshine and daisies; there's always tradeoffs in software development.
+
+## Challenges in the Hardware Renderer
+
+Earlier it was stated that the OpenGL hardware renderer was emulating the PICA200 by using the GPU instead of the CPU, and ... that's only partially true.
+As it stands, only a portion of the PICA200 emulation is running on the GPU; most of it is running on the CPU!
+To understand why, we need to dive a little deeper into the difference between CPUs and modern GPUs.
+
+{{< figure src="/images/entry/improvements-to-hardware-renderer/gpu_pipeline_before.png" 
+    title="The green shaded portions are what Citra emulates using the GPU. For a hardware renderer, Citra isn't using the GPU much!" 
+>}}
+
+As a general rule of thumb, CPUs are fast at computing general tasks, while GPUs are *blazing fast* at computing very specific tasks.
+Whenever the tasks the PICA200 performs matches well with tasks you can do on a GPU using OpenGL, everything is fast and everyone is happy!
+That said, we tend to run into edge cases that the PICA200 supports, but frankly, OpenGL is not well suited to support.
+This leads to cases where sometimes we just have to [live with minor inaccuracies as a tradeoff for speed](https://github.com/citra-emu/citra/pull/2697)
+
+{{< figure src="/images/entry/improvements-to-hardware-renderer/outline_bug.png" title="The infamous Pokémon outlines bug was one such example of something that OpenGL just doesn't handle well" >}}
+
+OpenGL is also great for emulator developers because it's a cross-platform standard for graphics, with support for all major desktop platforms.
+But because it's a standard, this means performance and features can vary widely between operating systems, graphics driver, and the physical graphics card.
+As you might have guessed, this leads to some [OS specific bugs that are very hard to track down](https://github.com/citra-emu/citra/issues/2416).
+In the linked issue, only on Mac OSX, Citra would leak memory from the hardware renderer.
+We traced it back to how textures were juggled between the 3DS memory and the host GPU, but we don't have many developers that use Mac, so we never did find the root cause.
+For a little bit of good news, **this is fixed in the latest nightly**, but only because the entire texture handling code was rewritten!
+
+## Moving Forward with the Hardware Renderer: Cleaning up Texture Forwarding
+
+Despite the issues mentioned above, OpenGL has been a fair choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making improving the renderer.
```

Comment:
`has been hard at work making improving the renderer.`

Should be `has been hard at work improving the renderer.`